src/engine/Input/PointerSystem.ts

```diff
@@ -0,0 +1,202 @@
+    return this.lastFrameEntityToPointers.has(entity.id) &&
+           this.lastFrameEntityToPointers.get(entity.id).includes(pointerId);
+  }
+
+  public entered(entity: Entity, pointerId: number) {
+    return this.entityCurrentlyUnderPointer(entity, pointerId) &&
+           !this.lastFrameEntityToPointers.has(entity.id);
+  }
+
+  public left(entity: Entity, pointerId: number) {
+    return !this.currentFrameEntityToPointers.has(entity.id) &&
+            this.entityWasUnderPointer(entity, pointerId);
+  }
+
+  public addPointerToEntity(entity: Entity, pointerId: number) {
+    if (!this.currentFrameEntityToPointers.has(entity.id)) {
+      this.currentFrameEntityToPointers.set(entity.id, [pointerId]);
+      return;
+    }
+    const pointers = this.currentFrameEntityToPointers.get(entity.id);
+    this.currentFrameEntityToPointers.set(entity.id, pointers.concat(pointerId));
+  }
+
+  public update(entities: Entity[]): void {
+    // Locate all the pointer/entity mappings
+    this._processPointerToEntity(entities);
+
+    // Dispatch pointer events on entities
+    this._dispatchEvents(entities);
+
+    // Clear last frame's events
+    this._receiver.update();
+    this.lastFrameEntityToPointers.clear();
+    this.lastFrameEntityToPointers = new Map<number, number[]>(this.currentFrameEntityToPointers);
+    this.currentFrameEntityToPointers.clear();
+    this._receiver.clear();
+  }
+
+  private _processPointerToEntity(entities: Entity[]) {
+    let transform: TransformComponent;
+    let collider: ColliderComponent;
+    let graphics: GraphicsComponent;
+    let pointerConfig: PointerComponent;
+
+    // TODO probably a spatial partition optimization here to quickly query bounds for pointer
+    // Pre-process find entities under pointers
+    for (const entity of entities) {
+      // transform = entity.get(TransformComponent);
+      pointerConfig = entity.get(PointerComponent) ?? new PointerComponent;
+      // Check collider contains pointer
+      collider = entity.get(ColliderComponent);
+      if (collider && pointerConfig.useColliderShape) {
+        const geom = collider.get();
+        if (geom) {
+          for (const [pointerId, pos] of this._receiver.currentFramePointerPositions.entries()) {
+            if (geom.contains(pos)) {
+              this.addPointerToEntity(entity, pointerId);
+            }
+          }
+        }
+      }
+
+      // Check graphics contains pointer
+      graphics = entity.get(GraphicsComponent);
+      if (graphics && pointerConfig.useGraphicsBounds) {
+        const graphicBounds = graphics.localBounds.transform(transform.getGlobalMatrix());
+        for (const [pointerId, pos] of this._receiver.currentFramePointerPositions.entries()) {
+          if (graphicBounds.contains(pos)) {
+            this.addPointerToEntity(entity, pointerId);
+          }
+        }
+      }
+    }
+  }
+
+  private _dispatchEvents(entities: Entity[]) {
```

Comment:
wish this could be simpler, or maybe extracted into smaller private helpers

Reply:
Let me noodle, there may be some things that can be broken out